src/pipe_windows.rs

```diff
@@ -0,0 +1,204 @@
+//!     do_something_that_may_be_interrupted_by_ctrl_c();
+//!     let mut buff = [0];
+//! # if false {
+//!     read.read_exact(&mut buff)?;
+//! # }
+//!     println!("Happily terminating");
+//!     Ok(())
+//! }
+//! ```
+
+use std::io::Error;
+use std::os::windows::prelude::*;
+use std::ptr;
+
+use libc::c_int;
+use winapi::um::fileapi::WriteFile;
+use winapi::um::winnt::HANDLE;
+use winapi::um::winsock2::{send, SOCKET};
+
+use SigId;
+
+pub(crate) fn wake_handle(pipe: RawHandle) {
+    unsafe {
+        let mut num_written = 0;
+        let _success = WriteFile(
+            pipe as HANDLE,
+            b"X" as *const _ as *const _,
+            1,
+            &mut num_written,
+            ptr::null_mut(),
+        );
+    }
+}
+
+pub(crate) fn wake_socket(pipe: RawSocket) {
+    unsafe {
+        send(pipe as SOCKET, b"X" as *const _ as *const _, 1, 0);
+    }
+}
+
+/// Registers a write to a self-pipe whenever there's the signal.
+///
+/// In this case, the pipe is taken as the `RawHandle`. It is still the caller's responsibility to
+/// close it.
+pub fn register_handle_raw(signal: c_int, pipe: RawHandle) -> Result<SigId, Error> {
+    struct SendHandle(RawHandle);
+    unsafe impl Send for SendHandle {}
+    unsafe impl Sync for SendHandle {}
+    let pipe = SendHandle(pipe);
+    unsafe { ::register(signal, move || wake_handle(pipe.0)) }
+}
+
+/// Registers a write to a self-pipe whenever there's the signal.
+///
+/// In this case, the pipe is taken as the `RawSocket`. It is still the caller's responsibility to
+/// close it.
+pub fn register_socket_raw(signal: c_int, pipe: RawSocket) -> Result<SigId, Error> {
+    struct SendSocket(RawSocket);
+    unsafe impl Send for SendSocket {}
+    unsafe impl Sync for SendSocket {}
+    let pipe = SendSocket(pipe);
+    unsafe { ::register(signal, move || wake_socket(pipe.0)) }
+}
+
+/// Registers a write to a self-pipe whenever there's the signal.
+///
+/// The ownership of pipe is taken and will be closed whenever the created action is unregistered.
+///
+/// Note that if you want to register the same pipe for multiple signals, there's `try_clone`
+/// method on many unix socket primitives.
+pub fn register_handle<P>(signal: c_int, pipe: P) -> Result<SigId, Error>
+where
+    P: AsRawHandle + Send + Sync + 'static,
+{
+    let action = move || wake_handle(pipe.as_raw_handle());
+    unsafe { ::register(signal, action) }
+}
+
+/// Registers a write to a self-pipe whenever there's the signal.
+///
+/// The ownership of pipe is taken and will be closed whenever the created action is unregistered.
+///
+/// Note that if you want to register the same pipe for multiple signals, there's `try_clone`
+/// method on many unix socket primitives.
+pub fn register_socket<P>(signal: c_int, pipe: P) -> Result<SigId, Error>
+where
+    P: AsRawSocket + Send + Sync + 'static,
+{
+    let action = move || wake_socket(pipe.as_raw_socket());
+    unsafe { ::register(signal, action) }
+}
```

Comment:
It seems a bit unfortunate that there are two sets of register functions here, one for handles, another for sockets. If you adopt the above suggestion about abstracting the differences away by a custom trait, could the AsPipe trait be implemented for both sets at once?

Reply:
I guess not that easy due to coherence.